lecture-notes/2-vcs.md

```diff
@@ -923,13 +921,129 @@ quote> not a number, or at least did not start with a digit.'
 
 ### How Git helps you
 
-TODO: `git log` is nice, but do you really care that much about simply viewing
-your old commit messages?
+The example above may have left you confused about the exact commits it actually
+made. Although running the commands yourself is generally less confusing than
+reading a transcript of them, it can still be hard to remember what you've
+committed and when.
+
+Fortunately, the entire purpose of Git is to make it so you don't have to
+remember! At long last, it's time to talk about all the Git subcommands that let
+you do interesting things based on the commits you've made. The next lecture and
+a half will be a whirlwind tour of some of those subcommands.
+
+#### Viewing history
+
+The most basic subcommand to inspect your commit history is `git log`, which lists every commit in reverse chronological order:
+
+```
+$ git log
+commit 6fd624ec083a21b76a0879697974ccc3820e14e8 (HEAD -> main)
+Author: Thomas Hebb <tommyhebb@gmail.com>
+Date:   Wed Jan 31 21:23:10 2024 -0500
+
+    Detect errors when parsing numbers
+
+    strtol can signal error via errno, but it can also report when it
+    stopped parsing input. This is helpful for determining if the input was
+    not a number, or at least did not start with a digit.
+
+commit 8ad441d1469c3e23bd7a261f9145f64179364c7c
+Author: Thomas Hebb <tommyhebb@gmail.com>
+Date:   Wed Jan 31 21:17:45 2024 -0500
+
+    Support the modulo operator
+
+commit 1838eeae89ba4cc4068bff8a1c99c0a9b0853506
+Author: Thomas Hebb <tommyhebb@gmail.com>
+Date:   Wed Jan 31 21:07:08 2024 -0500
+
+    Make main a simple calculator
+
+commit a83e7a6c4e9844b745a22450b69892d2292d8c7e
+Author: Thomas Hebb <tommyhebb@gmail.com>
+Date:   Tue Jan 30 22:59:25 2024 -0500
+
+    Add #include directives to fix compilation
+
+commit 3ba2a935126dafc8d4acb96df131bad26cb37396
+Author: Thomas Hebb <tommyhebb@gmail.com>
+Date:   Tue Jan 30 22:57:04 2024 -0500
+
+    Initial commit
+```
+
+TODO: description of how arguments are parsed for subcommands that take both an
+optional revision and an optional list of files (it's not pretty)
+
+Like most subcommands, you can alter the behavior of `git log` with various
+flags. `man git-log`, has a comprehensive list, but a couple notable ones are
+
+- **`--patch` (`-p`)**: Show each commit's changes, not just its metadata
+- **`--oneline`**: Show each commit in a condensed format on a single line
+
+```
```

Comment:
console

Reply:
make sure we reconcile commits as snapshots and displaying as diffs. maybe good to link jvns post